<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Source of the Rust file `bamnado/src/bam_utils.rs`."><title>bam_utils.rs - source</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2"href="../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../../static.files/rustdoc-ca0dd0c4.css"><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="bamnado" data-themes="" data-resource-suffix="" data-rustdoc-version="1.92.0 (ded5c06cf 2025-12-08)" data-channel="1.92.0" data-search-js="search-d69d8955.js" data-stringdex-js="stringdex-c3e638e9.js" data-settings-js="settings-c38705f0.js" ><script src="../../static.files/storage-e2aeef58.js"></script><script defer src="../../static.files/src-script-813739b1.js"></script><script defer src="../../src-files.js"></script><script defer src="../../static.files/main-ce535bd0.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-263c88ec.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-eab170b8.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-044be391.svg"></head><body class="rustdoc src"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="src-sidebar-title"><h2>Files</h2></div></nav><div class="sidebar-resizer" title="Drag to resize sidebar"></div><main><section id="main-content" class="content"><div class="main-heading"><h1><div class="sub-heading">bamnado/</div>bam_utils.rs</h1><rustdoc-toolbar></rustdoc-toolbar></div><div class="example-wrap digits-3"><pre class="rust"><code><a href=#1 id=1 data-nosnippet>1</a><span class="doccomment">//! # BAM Utilities Module
<a href=#2 id=2 data-nosnippet>2</a>//!
<a href=#3 id=3 data-nosnippet>3</a>//! This module contains a collection of utility functions and data structures for working with
<a href=#4 id=4 data-nosnippet>4</a>//! BAM files. It includes tools for:
<a href=#5 id=5 data-nosnippet>5</a>//! *   Parsing and handling BAM headers and indexes.
<a href=#6 id=6 data-nosnippet>6</a>//! *   Managing cell barcodes and strand information.
<a href=#7 id=7 data-nosnippet>7</a>//! *   Calculating basic statistics from BAM files.
<a href=#8 id=8 data-nosnippet>8</a>//! *   Helper types for genomic intervals and coordinate systems.
<a href=#9 id=9 data-nosnippet>9</a>
<a href=#10 id=10 data-nosnippet>10</a></span><span class="kw">use </span>ahash::{HashMap, HashSet};
<a href=#11 id=11 data-nosnippet>11</a><span class="kw">use </span>anyhow::{Context, <span class="prelude-ty">Result</span>};
<a href=#12 id=12 data-nosnippet>12</a><span class="kw">use </span>bio_types::annot::contig::Contig;
<a href=#13 id=13 data-nosnippet>13</a><span class="kw">use </span>bio_types::strand::ReqStrand;
<a href=#14 id=14 data-nosnippet>14</a><span class="kw">use </span>log::debug;
<a href=#15 id=15 data-nosnippet>15</a>
<a href=#16 id=16 data-nosnippet>16</a><span class="kw">use </span>bio_types::strand::Strand <span class="kw">as </span>BioStrand;
<a href=#17 id=17 data-nosnippet>17</a><span class="kw">use </span>noodles::core::{Position, Region};
<a href=#18 id=18 data-nosnippet>18</a><span class="kw">use </span>noodles::{bam, bed, sam};
<a href=#19 id=19 data-nosnippet>19</a><span class="kw">use </span>polars::prelude::<span class="kw-2">*</span>;
<a href=#20 id=20 data-nosnippet>20</a><span class="kw">use </span>rust_lapper::Interval;
<a href=#21 id=21 data-nosnippet>21</a><span class="kw">use </span>rust_lapper::Lapper;
<a href=#22 id=22 data-nosnippet>22</a><span class="kw">use </span>std::ops::Bound;
<a href=#23 id=23 data-nosnippet>23</a><span class="kw">use </span>std::path::{Path, PathBuf};
<a href=#24 id=24 data-nosnippet>24</a><span class="kw">use </span>std::str::FromStr;
<a href=#25 id=25 data-nosnippet>25</a>
<a href=#26 id=26 data-nosnippet>26</a><span class="doccomment">/// The cell barcode tag (CB).
<a href=#27 id=27 data-nosnippet>27</a></span><span class="kw">pub const </span>CB: [u8; <span class="number">2</span>] = [<span class="string">b'C'</span>, <span class="string">b'B'</span>];
<a href=#28 id=28 data-nosnippet>28</a><span class="doccomment">/// Type alias for an interval with a `u32` value.
<a href=#29 id=29 data-nosnippet>29</a></span><span class="kw">pub type </span>Iv = Interval&lt;usize, u32&gt;;
<a href=#30 id=30 data-nosnippet>30</a>
<a href=#31 id=31 data-nosnippet>31</a><span class="doccomment">/// Represents the strand of a genomic feature.
<a href=#32 id=32 data-nosnippet>32</a></span><span class="attr">#[derive(Debug, Clone, PartialEq, Eq, Copy)]
<a href=#33 id=33 data-nosnippet>33</a></span><span class="kw">pub enum </span>Strand {
<a href=#34 id=34 data-nosnippet>34</a>    <span class="doccomment">/// Forward strand (+).
<a href=#35 id=35 data-nosnippet>35</a>    </span>Forward,
<a href=#36 id=36 data-nosnippet>36</a>    <span class="doccomment">/// Reverse strand (-).
<a href=#37 id=37 data-nosnippet>37</a>    </span>Reverse,
<a href=#38 id=38 data-nosnippet>38</a>    <span class="doccomment">/// Both strands (unstranded).
<a href=#39 id=39 data-nosnippet>39</a>    </span>Both,
<a href=#40 id=40 data-nosnippet>40</a>}
<a href=#41 id=41 data-nosnippet>41</a><span class="kw">impl </span>FromStr <span class="kw">for </span>Strand {
<a href=#42 id=42 data-nosnippet>42</a>    <span class="kw">type </span><span class="prelude-val">Err </span>= String;
<a href=#43 id=43 data-nosnippet>43</a>
<a href=#44 id=44 data-nosnippet>44</a>    <span class="kw">fn </span>from_str(s: <span class="kw-2">&amp;</span>str) -&gt; <span class="prelude-ty">Result</span>&lt;<span class="self">Self</span>, <span class="self">Self</span>::Err&gt; {
<a href=#45 id=45 data-nosnippet>45</a>        <span class="kw">match </span>s.to_lowercase().as_str() {
<a href=#46 id=46 data-nosnippet>46</a>            <span class="string">"forward" </span>=&gt; <span class="prelude-val">Ok</span>(Strand::Forward),
<a href=#47 id=47 data-nosnippet>47</a>            <span class="string">"reverse" </span>=&gt; <span class="prelude-val">Ok</span>(Strand::Reverse),
<a href=#48 id=48 data-nosnippet>48</a>            <span class="string">"both" </span>=&gt; <span class="prelude-val">Ok</span>(Strand::Both),
<a href=#49 id=49 data-nosnippet>49</a>            <span class="kw">_ </span>=&gt; <span class="prelude-val">Err</span>(<span class="macro">format!</span>(<span class="string">"Invalid strand value: {s}"</span>)),
<a href=#50 id=50 data-nosnippet>50</a>        }
<a href=#51 id=51 data-nosnippet>51</a>    }
<a href=#52 id=52 data-nosnippet>52</a>}
<a href=#53 id=53 data-nosnippet>53</a><span class="kw">impl </span>std::fmt::Display <span class="kw">for </span>Strand {
<a href=#54 id=54 data-nosnippet>54</a>    <span class="kw">fn </span>fmt(<span class="kw-2">&amp;</span><span class="self">self</span>, f: <span class="kw-2">&amp;mut </span>std::fmt::Formatter&lt;<span class="lifetime">'_</span>&gt;) -&gt; std::fmt::Result {
<a href=#55 id=55 data-nosnippet>55</a>        <span class="kw">match </span><span class="self">self </span>{
<a href=#56 id=56 data-nosnippet>56</a>            Strand::Forward =&gt; <span class="macro">write!</span>(f, <span class="string">"forward"</span>),
<a href=#57 id=57 data-nosnippet>57</a>            Strand::Reverse =&gt; <span class="macro">write!</span>(f, <span class="string">"reverse"</span>),
<a href=#58 id=58 data-nosnippet>58</a>            Strand::Both =&gt; <span class="macro">write!</span>(f, <span class="string">"both"</span>),
<a href=#59 id=59 data-nosnippet>59</a>        }
<a href=#60 id=60 data-nosnippet>60</a>    }
<a href=#61 id=61 data-nosnippet>61</a>}
<a href=#62 id=62 data-nosnippet>62</a>
<a href=#63 id=63 data-nosnippet>63</a><span class="kw">impl </span>From&lt;Strand&gt; <span class="kw">for </span>BioStrand {
<a href=#64 id=64 data-nosnippet>64</a>    <span class="kw">fn </span>from(val: Strand) -&gt; <span class="self">Self </span>{
<a href=#65 id=65 data-nosnippet>65</a>        <span class="kw">match </span>val {
<a href=#66 id=66 data-nosnippet>66</a>            Strand::Forward =&gt; BioStrand::Forward,
<a href=#67 id=67 data-nosnippet>67</a>            Strand::Reverse =&gt; BioStrand::Reverse,
<a href=#68 id=68 data-nosnippet>68</a>            Strand::Both =&gt; BioStrand::Unknown,
<a href=#69 id=69 data-nosnippet>69</a>        }
<a href=#70 id=70 data-nosnippet>70</a>    }
<a href=#71 id=71 data-nosnippet>71</a>}
<a href=#72 id=72 data-nosnippet>72</a>
<a href=#73 id=73 data-nosnippet>73</a><span class="doccomment">/// Creates a progress bar with a standard style.
<a href=#74 id=74 data-nosnippet>74</a>///
<a href=#75 id=75 data-nosnippet>75</a>/// # Arguments
<a href=#76 id=76 data-nosnippet>76</a>///
<a href=#77 id=77 data-nosnippet>77</a>/// * `length` - The total number of items to process.
<a href=#78 id=78 data-nosnippet>78</a>/// * `message` - The message to display next to the progress bar.
<a href=#79 id=79 data-nosnippet>79</a></span><span class="kw">pub fn </span>progress_bar(length: u64, message: String) -&gt; indicatif::ProgressBar {
<a href=#80 id=80 data-nosnippet>80</a>    <span class="comment">// Progress bar
<a href=#81 id=81 data-nosnippet>81</a>    </span><span class="kw">let </span>progress_style = indicatif::ProgressStyle::with_template(
<a href=#82 id=82 data-nosnippet>82</a>        <span class="string">"{spinner:.green} [{elapsed_precise}] [{bar:40.cyan/blue}] {pos}/{len} ({eta}) {msg}"</span>,
<a href=#83 id=83 data-nosnippet>83</a>    )
<a href=#84 id=84 data-nosnippet>84</a>    .unwrap()
<a href=#85 id=85 data-nosnippet>85</a>    .progress_chars(<span class="string">"##-"</span>);
<a href=#86 id=86 data-nosnippet>86</a>
<a href=#87 id=87 data-nosnippet>87</a>    indicatif::ProgressBar::new(length)
<a href=#88 id=88 data-nosnippet>88</a>        .with_style(progress_style)
<a href=#89 id=89 data-nosnippet>89</a>        .with_message(message)
<a href=#90 id=90 data-nosnippet>90</a>}
<a href=#91 id=91 data-nosnippet>91</a>
<a href=#92 id=92 data-nosnippet>92</a><span class="doccomment">/// A collection of unique cell barcodes.
<a href=#93 id=93 data-nosnippet>93</a></span><span class="kw">pub struct </span>CellBarcodes {
<a href=#94 id=94 data-nosnippet>94</a>    barcodes: HashSet&lt;String&gt;,
<a href=#95 id=95 data-nosnippet>95</a>}
<a href=#96 id=96 data-nosnippet>96</a>
<a href=#97 id=97 data-nosnippet>97</a><span class="kw">impl </span>Default <span class="kw">for </span>CellBarcodes {
<a href=#98 id=98 data-nosnippet>98</a>    <span class="kw">fn </span>default() -&gt; <span class="self">Self </span>{
<a href=#99 id=99 data-nosnippet>99</a>        <span class="self">Self</span>::new()
<a href=#100 id=100 data-nosnippet>100</a>    }
<a href=#101 id=101 data-nosnippet>101</a>}
<a href=#102 id=102 data-nosnippet>102</a>
<a href=#103 id=103 data-nosnippet>103</a><span class="kw">impl </span>CellBarcodes {
<a href=#104 id=104 data-nosnippet>104</a>    <span class="doccomment">/// Creates a new empty `CellBarcodes` collection.
<a href=#105 id=105 data-nosnippet>105</a>    </span><span class="kw">pub fn </span>new() -&gt; <span class="self">Self </span>{
<a href=#106 id=106 data-nosnippet>106</a>        <span class="self">Self </span>{
<a href=#107 id=107 data-nosnippet>107</a>            barcodes: HashSet::default(),
<a href=#108 id=108 data-nosnippet>108</a>        }
<a href=#109 id=109 data-nosnippet>109</a>    }
<a href=#110 id=110 data-nosnippet>110</a>
<a href=#111 id=111 data-nosnippet>111</a>    <span class="doccomment">/// Returns a copy of the set of barcodes.
<a href=#112 id=112 data-nosnippet>112</a>    </span><span class="kw">pub fn </span>barcodes(<span class="kw-2">&amp;</span><span class="self">self</span>) -&gt; HashSet&lt;String&gt; {
<a href=#113 id=113 data-nosnippet>113</a>        <span class="self">self</span>.barcodes.clone()
<a href=#114 id=114 data-nosnippet>114</a>    }
<a href=#115 id=115 data-nosnippet>115</a>
<a href=#116 id=116 data-nosnippet>116</a>    <span class="doccomment">/// Returns `true` if the collection is empty.
<a href=#117 id=117 data-nosnippet>117</a>    </span><span class="kw">pub fn </span>is_empty(<span class="kw-2">&amp;</span><span class="self">self</span>) -&gt; bool {
<a href=#118 id=118 data-nosnippet>118</a>        <span class="self">self</span>.barcodes.is_empty()
<a href=#119 id=119 data-nosnippet>119</a>    }
<a href=#120 id=120 data-nosnippet>120</a>
<a href=#121 id=121 data-nosnippet>121</a>    <span class="doccomment">/// Loads cell barcodes from a CSV file.
<a href=#122 id=122 data-nosnippet>122</a>    ///
<a href=#123 id=123 data-nosnippet>123</a>    /// The CSV file must have a header with a column named "barcode".
<a href=#124 id=124 data-nosnippet>124</a>    </span><span class="kw">pub fn </span>from_csv(file_path: <span class="kw-2">&amp;</span>PathBuf) -&gt; <span class="prelude-ty">Result</span>&lt;<span class="self">Self</span>&gt; {
<a href=#125 id=125 data-nosnippet>125</a>        <span class="kw">let </span>path = Path::new(file_path).to_path_buf();
<a href=#126 id=126 data-nosnippet>126</a>
<a href=#127 id=127 data-nosnippet>127</a>        <span class="kw">let </span>df = CsvReadOptions::default()
<a href=#128 id=128 data-nosnippet>128</a>            .with_has_header(<span class="bool-val">true</span>)
<a href=#129 id=129 data-nosnippet>129</a>            .try_into_reader_with_file_path(<span class="prelude-val">Some</span>(path))<span class="question-mark">?
<a href=#130 id=130 data-nosnippet>130</a>            </span>.finish()<span class="question-mark">?</span>;
<a href=#131 id=131 data-nosnippet>131</a>        <span class="kw">let </span><span class="kw-2">mut </span>barcodes = HashSet::default();
<a href=#132 id=132 data-nosnippet>132</a>
<a href=#133 id=133 data-nosnippet>133</a>        <span class="kw">for </span>barcode <span class="kw">in </span>df.column(<span class="string">"barcode"</span>).unwrap().str().unwrap() {
<a href=#134 id=134 data-nosnippet>134</a>            <span class="kw">let </span>barcode = barcode.unwrap().to_string();
<a href=#135 id=135 data-nosnippet>135</a>            barcodes.insert(barcode);
<a href=#136 id=136 data-nosnippet>136</a>        }
<a href=#137 id=137 data-nosnippet>137</a>
<a href=#138 id=138 data-nosnippet>138</a>        <span class="macro">println!</span>(<span class="string">"Number of barcodes: {}"</span>, barcodes.len());
<a href=#139 id=139 data-nosnippet>139</a>        <span class="macro">println!</span>(
<a href=#140 id=140 data-nosnippet>140</a>            <span class="string">"First 10 barcodes: {:?}"</span>,
<a href=#141 id=141 data-nosnippet>141</a>            barcodes.iter().take(<span class="number">10</span>).collect::&lt;Vec&lt;<span class="kw">_</span>&gt;&gt;()
<a href=#142 id=142 data-nosnippet>142</a>        );
<a href=#143 id=143 data-nosnippet>143</a>
<a href=#144 id=144 data-nosnippet>144</a>        <span class="prelude-val">Ok</span>(<span class="self">Self </span>{ barcodes })
<a href=#145 id=145 data-nosnippet>145</a>    }
<a href=#146 id=146 data-nosnippet>146</a>}
<a href=#147 id=147 data-nosnippet>147</a>
<a href=#148 id=148 data-nosnippet>148</a><span class="doccomment">/// A collection of sets of cell barcodes, where each set represents a group.
<a href=#149 id=149 data-nosnippet>149</a></span><span class="kw">pub struct </span>CellBarcodesMulti {
<a href=#150 id=150 data-nosnippet>150</a>    barcodes: Vec&lt;HashSet&lt;String&gt;&gt;,
<a href=#151 id=151 data-nosnippet>151</a>}
<a href=#152 id=152 data-nosnippet>152</a>
<a href=#153 id=153 data-nosnippet>153</a><span class="kw">impl </span>Default <span class="kw">for </span>CellBarcodesMulti {
<a href=#154 id=154 data-nosnippet>154</a>    <span class="kw">fn </span>default() -&gt; <span class="self">Self </span>{
<a href=#155 id=155 data-nosnippet>155</a>        <span class="self">Self</span>::new()
<a href=#156 id=156 data-nosnippet>156</a>    }
<a href=#157 id=157 data-nosnippet>157</a>}
<a href=#158 id=158 data-nosnippet>158</a>
<a href=#159 id=159 data-nosnippet>159</a><span class="kw">impl </span>CellBarcodesMulti {
<a href=#160 id=160 data-nosnippet>160</a>    <span class="doccomment">/// Creates a new empty `CellBarcodesMulti` collection.
<a href=#161 id=161 data-nosnippet>161</a>    </span><span class="kw">pub fn </span>new() -&gt; <span class="self">Self </span>{
<a href=#162 id=162 data-nosnippet>162</a>        <span class="self">Self </span>{
<a href=#163 id=163 data-nosnippet>163</a>            barcodes: Vec::new(),
<a href=#164 id=164 data-nosnippet>164</a>        }
<a href=#165 id=165 data-nosnippet>165</a>    }
<a href=#166 id=166 data-nosnippet>166</a>
<a href=#167 id=167 data-nosnippet>167</a>    <span class="doccomment">/// Returns a copy of the vector of barcode sets.
<a href=#168 id=168 data-nosnippet>168</a>    </span><span class="kw">pub fn </span>barcodes(<span class="kw-2">&amp;</span><span class="self">self</span>) -&gt; Vec&lt;HashSet&lt;String&gt;&gt; {
<a href=#169 id=169 data-nosnippet>169</a>        <span class="self">self</span>.barcodes.clone()
<a href=#170 id=170 data-nosnippet>170</a>    }
<a href=#171 id=171 data-nosnippet>171</a>
<a href=#172 id=172 data-nosnippet>172</a>    <span class="doccomment">/// Returns `true` if the collection is empty.
<a href=#173 id=173 data-nosnippet>173</a>    </span><span class="kw">pub fn </span>is_empty(<span class="kw-2">&amp;</span><span class="self">self</span>) -&gt; bool {
<a href=#174 id=174 data-nosnippet>174</a>        <span class="self">self</span>.barcodes.is_empty()
<a href=#175 id=175 data-nosnippet>175</a>    }
<a href=#176 id=176 data-nosnippet>176</a>
<a href=#177 id=177 data-nosnippet>177</a>    <span class="doccomment">/// Loads multiple sets of barcodes from a CSV file.
<a href=#178 id=178 data-nosnippet>178</a>    ///
<a href=#179 id=179 data-nosnippet>179</a>    /// Each column in the CSV file is treated as a separate set of barcodes.
<a href=#180 id=180 data-nosnippet>180</a>    </span><span class="kw">pub fn </span>from_csv(file_path: <span class="kw-2">&amp;</span>PathBuf) -&gt; <span class="prelude-ty">Result</span>&lt;<span class="self">Self</span>&gt; {
<a href=#181 id=181 data-nosnippet>181</a>        <span class="kw">let </span>path = Path::new(file_path).to_path_buf();
<a href=#182 id=182 data-nosnippet>182</a>
<a href=#183 id=183 data-nosnippet>183</a>        <span class="kw">let </span>df = CsvReadOptions::default()
<a href=#184 id=184 data-nosnippet>184</a>            .with_has_header(<span class="bool-val">true</span>)
<a href=#185 id=185 data-nosnippet>185</a>            .try_into_reader_with_file_path(<span class="prelude-val">Some</span>(path))<span class="question-mark">?
<a href=#186 id=186 data-nosnippet>186</a>            </span>.finish()<span class="question-mark">?</span>;
<a href=#187 id=187 data-nosnippet>187</a>
<a href=#188 id=188 data-nosnippet>188</a>        <span class="kw">let </span><span class="kw-2">mut </span>barcodes = Vec::new();
<a href=#189 id=189 data-nosnippet>189</a>
<a href=#190 id=190 data-nosnippet>190</a>        <span class="kw">for </span>column <span class="kw">in </span>df.iter() {
<a href=#191 id=191 data-nosnippet>191</a>            <span class="kw">let </span>barcode = column.str().unwrap();
<a href=#192 id=192 data-nosnippet>192</a>
<a href=#193 id=193 data-nosnippet>193</a>            <span class="kw">let </span>bc = barcode
<a href=#194 id=194 data-nosnippet>194</a>                .iter()
<a href=#195 id=195 data-nosnippet>195</a>                .filter_map(|x| x.map(|x| x.to_string()))
<a href=#196 id=196 data-nosnippet>196</a>                .collect::&lt;HashSet&lt;String&gt;&gt;();
<a href=#197 id=197 data-nosnippet>197</a>
<a href=#198 id=198 data-nosnippet>198</a>            barcodes.push(bc);
<a href=#199 id=199 data-nosnippet>199</a>        }
<a href=#200 id=200 data-nosnippet>200</a>
<a href=#201 id=201 data-nosnippet>201</a>        <span class="prelude-val">Ok</span>(<span class="self">Self </span>{ barcodes })
<a href=#202 id=202 data-nosnippet>202</a>    }
<a href=#203 id=203 data-nosnippet>203</a>}
<a href=#204 id=204 data-nosnippet>204</a>
<a href=#205 id=205 data-nosnippet>205</a><span class="attr">#[derive(Debug)]
<a href=#206 id=206 data-nosnippet>206</a>#[allow(dead_code)]
<a href=#207 id=207 data-nosnippet>207</a></span><span class="kw">struct </span>ChromosomeStats {
<a href=#208 id=208 data-nosnippet>208</a>    chrom: String,
<a href=#209 id=209 data-nosnippet>209</a>    length: u64,
<a href=#210 id=210 data-nosnippet>210</a>    mapped: u64,
<a href=#211 id=211 data-nosnippet>211</a>    unmapped: u64,
<a href=#212 id=212 data-nosnippet>212</a>}
<a href=#213 id=213 data-nosnippet>213</a>
<a href=#214 id=214 data-nosnippet>214</a><span class="doccomment">/// Reads the header of a BAM file.
<a href=#215 id=215 data-nosnippet>215</a>///
<a href=#216 id=216 data-nosnippet>216</a>/// If the `noodles` crate fails to read the header, it falls back to `samtools`.
<a href=#217 id=217 data-nosnippet>217</a>/// This is useful for handling BAM files with slightly non-standard headers (e.g., from CellRanger).
<a href=#218 id=218 data-nosnippet>218</a></span><span class="kw">pub fn </span>bam_header(file_path: PathBuf) -&gt; <span class="prelude-ty">Result</span>&lt;sam::Header&gt; {
<a href=#219 id=219 data-nosnippet>219</a>    <span class="comment">// Read the header of a BAM file
<a href=#220 id=220 data-nosnippet>220</a>    // If the noodles crate fails to read the header, we fall back to samtools
<a href=#221 id=221 data-nosnippet>221</a>    // This is a bit of a hack but it works for now
<a href=#222 id=222 data-nosnippet>222</a>    // The noodles crate is more strict about the header format than samtools
<a href=#223 id=223 data-nosnippet>223</a>
<a href=#224 id=224 data-nosnippet>224</a>    // Check that the file exists
<a href=#225 id=225 data-nosnippet>225</a>    </span><span class="kw">if </span>!file_path.exists() {
<a href=#226 id=226 data-nosnippet>226</a>        <span class="kw">return </span><span class="prelude-val">Err</span>(anyhow::Error::from(std::io::Error::new(
<a href=#227 id=227 data-nosnippet>227</a>            std::io::ErrorKind::NotFound,
<a href=#228 id=228 data-nosnippet>228</a>            <span class="macro">format!</span>(<span class="string">"File not found: {}"</span>, file_path.display()),
<a href=#229 id=229 data-nosnippet>229</a>        )));
<a href=#230 id=230 data-nosnippet>230</a>    };
<a href=#231 id=231 data-nosnippet>231</a>
<a href=#232 id=232 data-nosnippet>232</a>    <span class="kw">let </span><span class="kw-2">mut </span>reader = bam::io::indexed_reader::Builder::default()
<a href=#233 id=233 data-nosnippet>233</a>        .build_from_path(file_path.clone())
<a href=#234 id=234 data-nosnippet>234</a>        .expect(<span class="string">"Failed to open file"</span>);
<a href=#235 id=235 data-nosnippet>235</a>
<a href=#236 id=236 data-nosnippet>236</a>    <span class="kw">let </span>header = <span class="kw">match </span>reader.read_header() {
<a href=#237 id=237 data-nosnippet>237</a>        std::result::Result::Ok(header) =&gt; header,
<a href=#238 id=238 data-nosnippet>238</a>        <span class="prelude-val">Err</span>(e) =&gt; {
<a href=#239 id=239 data-nosnippet>239</a>            <span class="macro">debug!</span>(<span class="string">"Failed to read header using noodels falling back to samtools: {e}"</span>);
<a href=#240 id=240 data-nosnippet>240</a>
<a href=#241 id=241 data-nosnippet>241</a>            <span class="kw">let </span>header_samtools = std::process::Command::new(<span class="string">"samtools"</span>)
<a href=#242 id=242 data-nosnippet>242</a>                .arg(<span class="string">"view"</span>)
<a href=#243 id=243 data-nosnippet>243</a>                .arg(<span class="string">"-H"</span>)
<a href=#244 id=244 data-nosnippet>244</a>                .arg(file_path.clone())
<a href=#245 id=245 data-nosnippet>245</a>                .output()
<a href=#246 id=246 data-nosnippet>246</a>                .expect(<span class="string">"Failed to run samtools"</span>)
<a href=#247 id=247 data-nosnippet>247</a>                .stdout;
<a href=#248 id=248 data-nosnippet>248</a>
<a href=#249 id=249 data-nosnippet>249</a>            <span class="kw">let </span>header_str =
<a href=#250 id=250 data-nosnippet>250</a>                String::from_utf8(header_samtools).expect(<span class="string">"Failed to convert header to string"</span>);
<a href=#251 id=251 data-nosnippet>251</a>
<a href=#252 id=252 data-nosnippet>252</a>            <span class="comment">// Slight hack here for CellRanger BAM files that are missing the version info
<a href=#253 id=253 data-nosnippet>253</a>            </span><span class="kw">let </span>header_string =
<a href=#254 id=254 data-nosnippet>254</a>                header_str.replace(<span class="string">"@HD\tSO:coordinate\n"</span>, <span class="string">"@HD\tVN:1.6\tSO:coordinate\n"</span>);
<a href=#255 id=255 data-nosnippet>255</a>            <span class="kw">let </span>header_str = header_string.as_bytes();
<a href=#256 id=256 data-nosnippet>256</a>            <span class="kw">let </span><span class="kw-2">mut </span>reader = sam::io::Reader::new(header_str);
<a href=#257 id=257 data-nosnippet>257</a>            reader
<a href=#258 id=258 data-nosnippet>258</a>                .read_header()
<a href=#259 id=259 data-nosnippet>259</a>                .expect(<span class="string">"Failed to read header with samtools"</span>)
<a href=#260 id=260 data-nosnippet>260</a>        }
<a href=#261 id=261 data-nosnippet>261</a>    };
<a href=#262 id=262 data-nosnippet>262</a>    <span class="prelude-val">Ok</span>(header)
<a href=#263 id=263 data-nosnippet>263</a>}
<a href=#264 id=264 data-nosnippet>264</a>
<a href=#265 id=265 data-nosnippet>265</a><span class="doccomment">/// Stores statistics and metadata about a BAM file.
<a href=#266 id=266 data-nosnippet>266</a></span><span class="kw">pub struct </span>BamStats {
<a href=#267 id=267 data-nosnippet>267</a>    <span class="comment">// BAM file stats
<a href=#268 id=268 data-nosnippet>268</a>
<a href=#269 id=269 data-nosnippet>269</a>    // File path
<a href=#270 id=270 data-nosnippet>270</a>    </span><span class="attr">#[allow(dead_code)]
<a href=#271 id=271 data-nosnippet>271</a>    </span>file_path: PathBuf,
<a href=#272 id=272 data-nosnippet>272</a>
<a href=#273 id=273 data-nosnippet>273</a>    <span class="comment">// Header
<a href=#274 id=274 data-nosnippet>274</a>    </span>header: sam::Header,
<a href=#275 id=275 data-nosnippet>275</a>
<a href=#276 id=276 data-nosnippet>276</a>    <span class="comment">// Contigs present in the BAM file
<a href=#277 id=277 data-nosnippet>277</a>    </span><span class="attr">#[allow(dead_code)]
<a href=#278 id=278 data-nosnippet>278</a>    </span>contigs: Vec&lt;Contig&lt;String, ReqStrand&gt;&gt;,
<a href=#279 id=279 data-nosnippet>279</a>
<a href=#280 id=280 data-nosnippet>280</a>    <span class="comment">// Chromosome stats
<a href=#281 id=281 data-nosnippet>281</a>    </span>chrom_stats: HashMap&lt;String, ChromosomeStats&gt;,
<a href=#282 id=282 data-nosnippet>282</a>
<a href=#283 id=283 data-nosnippet>283</a>    <span class="comment">// Total number of reads
<a href=#284 id=284 data-nosnippet>284</a>    </span>n_reads: u64,
<a href=#285 id=285 data-nosnippet>285</a>    <span class="comment">// Number of reads mapped to the genome
<a href=#286 id=286 data-nosnippet>286</a>    </span>n_mapped: u64,
<a href=#287 id=287 data-nosnippet>287</a>    <span class="comment">// Number of reads unmapped to the genome
<a href=#288 id=288 data-nosnippet>288</a>    </span>n_unmapped: u64,
<a href=#289 id=289 data-nosnippet>289</a>}
<a href=#290 id=290 data-nosnippet>290</a>
<a href=#291 id=291 data-nosnippet>291</a><span class="kw">impl </span>BamStats {
<a href=#292 id=292 data-nosnippet>292</a>    <span class="doccomment">/// Creates a new `BamStats` instance from a BAM file path.
<a href=#293 id=293 data-nosnippet>293</a>    ///
<a href=#294 id=294 data-nosnippet>294</a>    /// This reads the header and index to calculate statistics.
<a href=#295 id=295 data-nosnippet>295</a>    </span><span class="kw">pub fn </span>new(file_path: PathBuf) -&gt; <span class="prelude-ty">Result</span>&lt;<span class="self">Self</span>&gt; {
<a href=#296 id=296 data-nosnippet>296</a>        <span class="comment">// Read the header of the BAM file
<a href=#297 id=297 data-nosnippet>297</a>        </span><span class="kw">let </span>header = bam_header(file_path.clone())<span class="question-mark">?</span>;
<a href=#298 id=298 data-nosnippet>298</a>
<a href=#299 id=299 data-nosnippet>299</a>        <span class="comment">// Get the contigs from the header
<a href=#300 id=300 data-nosnippet>300</a>        </span><span class="kw">let </span>contigs = header
<a href=#301 id=301 data-nosnippet>301</a>            .reference_sequences()
<a href=#302 id=302 data-nosnippet>302</a>            .iter()
<a href=#303 id=303 data-nosnippet>303</a>            .map(|(name, map)| {
<a href=#304 id=304 data-nosnippet>304</a>                <span class="kw">let </span>name = name.to_string();
<a href=#305 id=305 data-nosnippet>305</a>                <span class="kw">let </span>length = map.length().get();
<a href=#306 id=306 data-nosnippet>306</a>                Contig::new(name, <span class="number">1</span>, length, ReqStrand::Forward)
<a href=#307 id=307 data-nosnippet>307</a>            })
<a href=#308 id=308 data-nosnippet>308</a>            .collect();
<a href=#309 id=309 data-nosnippet>309</a>
<a href=#310 id=310 data-nosnippet>310</a>        <span class="comment">// Get the index of the BAM file
<a href=#311 id=311 data-nosnippet>311</a>        </span><span class="kw">let </span>bam_reader = bam::io::indexed_reader::Builder::default()
<a href=#312 id=312 data-nosnippet>312</a>            .build_from_path(file_path.clone())
<a href=#313 id=313 data-nosnippet>313</a>            .expect(<span class="string">"Failed to open file"</span>);
<a href=#314 id=314 data-nosnippet>314</a>        <span class="kw">let </span>index = bam_reader.index();
<a href=#315 id=315 data-nosnippet>315</a>
<a href=#316 id=316 data-nosnippet>316</a>        <span class="comment">// Get the chromosome stats
<a href=#317 id=317 data-nosnippet>317</a>        </span><span class="kw">let </span><span class="kw-2">mut </span>chrom_stats = HashMap::default();
<a href=#318 id=318 data-nosnippet>318</a>
<a href=#319 id=319 data-nosnippet>319</a>        <span class="kw">for </span>((reference_sequence_name_buf, reference_sequence), index_reference_sequence) <span class="kw">in </span>header
<a href=#320 id=320 data-nosnippet>320</a>            .reference_sequences()
<a href=#321 id=321 data-nosnippet>321</a>            .iter()
<a href=#322 id=322 data-nosnippet>322</a>            .zip(index.reference_sequences())
<a href=#323 id=323 data-nosnippet>323</a>        {
<a href=#324 id=324 data-nosnippet>324</a>            <span class="kw">let </span>reference_sequence_name = std::str::from_utf8(reference_sequence_name_buf)
<a href=#325 id=325 data-nosnippet>325</a>                .map_err(|e| std::io::Error::new(std::io::ErrorKind::InvalidInput, e))<span class="question-mark">?</span>;
<a href=#326 id=326 data-nosnippet>326</a>
<a href=#327 id=327 data-nosnippet>327</a>            <span class="kw">let </span>(mapped_record_count, unmapped_record_count) = index_reference_sequence
<a href=#328 id=328 data-nosnippet>328</a>                .metadata()
<a href=#329 id=329 data-nosnippet>329</a>                .map(|m| (m.mapped_record_count(), m.unmapped_record_count()))
<a href=#330 id=330 data-nosnippet>330</a>                .unwrap_or_default();
<a href=#331 id=331 data-nosnippet>331</a>
<a href=#332 id=332 data-nosnippet>332</a>            <span class="kw">let </span>stats = ChromosomeStats {
<a href=#333 id=333 data-nosnippet>333</a>                chrom: reference_sequence_name.to_string(),
<a href=#334 id=334 data-nosnippet>334</a>                length: usize::from(reference_sequence.length()) <span class="kw">as </span>u64,
<a href=#335 id=335 data-nosnippet>335</a>                mapped: mapped_record_count,
<a href=#336 id=336 data-nosnippet>336</a>                unmapped: unmapped_record_count,
<a href=#337 id=337 data-nosnippet>337</a>            };
<a href=#338 id=338 data-nosnippet>338</a>
<a href=#339 id=339 data-nosnippet>339</a>            chrom_stats.insert(reference_sequence_name.to_string(), stats);
<a href=#340 id=340 data-nosnippet>340</a>        }
<a href=#341 id=341 data-nosnippet>341</a>
<a href=#342 id=342 data-nosnippet>342</a>        <span class="kw">let </span><span class="kw-2">mut </span>unmapped_record_count = index.unplaced_unmapped_record_count().unwrap_or_default();
<a href=#343 id=343 data-nosnippet>343</a>
<a href=#344 id=344 data-nosnippet>344</a>        <span class="kw">let </span><span class="kw-2">mut </span>mapped_record_count = <span class="number">0</span>;
<a href=#345 id=345 data-nosnippet>345</a>
<a href=#346 id=346 data-nosnippet>346</a>        <span class="kw">for </span>(<span class="kw">_</span>, stats) <span class="kw">in </span>chrom_stats.iter() {
<a href=#347 id=347 data-nosnippet>347</a>            mapped_record_count += stats.mapped;
<a href=#348 id=348 data-nosnippet>348</a>            unmapped_record_count += stats.unmapped;
<a href=#349 id=349 data-nosnippet>349</a>        }
<a href=#350 id=350 data-nosnippet>350</a>
<a href=#351 id=351 data-nosnippet>351</a>        <span class="kw">let </span>n_reads = mapped_record_count + unmapped_record_count;
<a href=#352 id=352 data-nosnippet>352</a>
<a href=#353 id=353 data-nosnippet>353</a>        <span class="prelude-val">Ok</span>(<span class="self">Self </span>{
<a href=#354 id=354 data-nosnippet>354</a>            file_path,
<a href=#355 id=355 data-nosnippet>355</a>            header,
<a href=#356 id=356 data-nosnippet>356</a>            contigs,
<a href=#357 id=357 data-nosnippet>357</a>            chrom_stats,
<a href=#358 id=358 data-nosnippet>358</a>            n_reads,
<a href=#359 id=359 data-nosnippet>359</a>            n_mapped: mapped_record_count,
<a href=#360 id=360 data-nosnippet>360</a>            n_unmapped: unmapped_record_count,
<a href=#361 id=361 data-nosnippet>361</a>        })
<a href=#362 id=362 data-nosnippet>362</a>    }
<a href=#363 id=363 data-nosnippet>363</a>
<a href=#364 id=364 data-nosnippet>364</a>    <span class="doccomment">/// Estimates the optimal genome chunk length for parallel processing.
<a href=#365 id=365 data-nosnippet>365</a>    ///
<a href=#366 id=366 data-nosnippet>366</a>    /// The chunk length is calculated based on the genome size, number of mapped reads,
<a href=#367 id=367 data-nosnippet>367</a>    /// and the desired bin size.
<a href=#368 id=368 data-nosnippet>368</a>    </span><span class="kw">pub fn </span>estimate_genome_chunk_length(<span class="kw-2">&amp;</span><span class="self">self</span>, bin_size: u64) -&gt; <span class="prelude-ty">Result</span>&lt;u64&gt; {
<a href=#369 id=369 data-nosnippet>369</a>        <span class="comment">// genomeLength = sum(bamHandles[0].lengths)
<a href=#370 id=370 data-nosnippet>370</a>        // max_reads_per_bp = max([float(x) / genomeLength for x in mappedList])
<a href=#371 id=371 data-nosnippet>371</a>        // genomeChunkLength = int(min(5e6, int(2e6 / (max_reads_per_bp * len(bamHandles)))))
<a href=#372 id=372 data-nosnippet>372</a>        // genomeChunkLength -= genomeChunkLength % tile_size
<a href=#373 id=373 data-nosnippet>373</a>
<a href=#374 id=374 data-nosnippet>374</a>        </span><span class="kw">let </span>stats = <span class="kw-2">&amp;</span><span class="self">self</span>.chrom_stats;
<a href=#375 id=375 data-nosnippet>375</a>        <span class="kw">let </span>genome_length = stats.values().map(|x| x.length).sum::&lt;u64&gt;();
<a href=#376 id=376 data-nosnippet>376</a>        <span class="kw">let </span>max_reads_per_bp = <span class="self">self</span>.n_mapped <span class="kw">as </span>f64 / genome_length <span class="kw">as </span>f64;
<a href=#377 id=377 data-nosnippet>377</a>        <span class="kw">let </span>genome_chunk_length = <span class="number">5e6_f64</span>.min(<span class="number">2e6 </span>/ (max_reads_per_bp));
<a href=#378 id=378 data-nosnippet>378</a>
<a href=#379 id=379 data-nosnippet>379</a>        <span class="kw">let </span>correction = genome_chunk_length % bin_size <span class="kw">as </span>f64;
<a href=#380 id=380 data-nosnippet>380</a>        <span class="kw">let </span>genome_chunk_length = genome_chunk_length - correction;
<a href=#381 id=381 data-nosnippet>381</a>
<a href=#382 id=382 data-nosnippet>382</a>        <span class="prelude-val">Ok</span>(genome_chunk_length <span class="kw">as </span>u64)
<a href=#383 id=383 data-nosnippet>383</a>    }
<a href=#384 id=384 data-nosnippet>384</a>
<a href=#385 id=385 data-nosnippet>385</a>    <span class="doccomment">/// Splits the genome into chunks for parallel processing.
<a href=#386 id=386 data-nosnippet>386</a>    </span><span class="kw">pub fn </span>genome_chunks(<span class="kw-2">&amp;</span><span class="self">self</span>, bin_size: u64) -&gt; <span class="prelude-ty">Result</span>&lt;Vec&lt;Region&gt;&gt; {
<a href=#387 id=387 data-nosnippet>387</a>        <span class="kw">let </span>genome_chunk_length = <span class="self">self</span>.estimate_genome_chunk_length(bin_size)<span class="question-mark">?</span>;
<a href=#388 id=388 data-nosnippet>388</a>        <span class="kw">let </span>chrom_chunks = <span class="self">self
<a href=#389 id=389 data-nosnippet>389</a>            </span>.chrom_stats
<a href=#390 id=390 data-nosnippet>390</a>            .iter()
<a href=#391 id=391 data-nosnippet>391</a>            .flat_map(|(chrom, stats)| {
<a href=#392 id=392 data-nosnippet>392</a>                <span class="kw">let </span><span class="kw-2">mut </span>chunks = Vec::new();
<a href=#393 id=393 data-nosnippet>393</a>                <span class="kw">let </span><span class="kw-2">mut </span>chunk_start = <span class="number">1</span>;
<a href=#394 id=394 data-nosnippet>394</a>                <span class="kw">let </span>chrom_end = stats.length;
<a href=#395 id=395 data-nosnippet>395</a>
<a href=#396 id=396 data-nosnippet>396</a>                <span class="kw">while </span>chunk_start &lt;= chrom_end {
<a href=#397 id=397 data-nosnippet>397</a>                    <span class="comment">// Corrected to include the last position in the range
<a href=#398 id=398 data-nosnippet>398</a>                    </span><span class="kw">let </span>chunk_end = chunk_start + genome_chunk_length - <span class="number">1</span>; <span class="comment">// Adjust to ensure the chunk covers exactly genome_chunk_length positions
<a href=#399 id=399 data-nosnippet>399</a>                    </span><span class="kw">let </span>chunk_end = chunk_end.min(chrom_end); <span class="comment">// Ensure we do not exceed the chromosome length
<a href=#400 id=400 data-nosnippet>400</a>
<a href=#401 id=401 data-nosnippet>401</a>                    </span><span class="kw">let </span>start = Position::try_from(chunk_start <span class="kw">as </span>usize).unwrap();
<a href=#402 id=402 data-nosnippet>402</a>                    <span class="kw">let </span>end = Position::try_from(chunk_end <span class="kw">as </span>usize).unwrap();
<a href=#403 id=403 data-nosnippet>403</a>
<a href=#404 id=404 data-nosnippet>404</a>                    <span class="kw">let </span>region = Region::new(<span class="kw-2">&amp;*</span>chrom.clone(), start..=end);
<a href=#405 id=405 data-nosnippet>405</a>                    chunks.push(region);
<a href=#406 id=406 data-nosnippet>406</a>                    chunk_start = chunk_end + <span class="number">1</span>; <span class="comment">// Corrected to start the next chunk right after the current chunk ends
<a href=#407 id=407 data-nosnippet>407</a>                </span>}
<a href=#408 id=408 data-nosnippet>408</a>                chunks
<a href=#409 id=409 data-nosnippet>409</a>            })
<a href=#410 id=410 data-nosnippet>410</a>            .collect();
<a href=#411 id=411 data-nosnippet>411</a>
<a href=#412 id=412 data-nosnippet>412</a>        <span class="prelude-val">Ok</span>(chrom_chunks)
<a href=#413 id=413 data-nosnippet>413</a>    }
<a href=#414 id=414 data-nosnippet>414</a>
<a href=#415 id=415 data-nosnippet>415</a>    <span class="doccomment">/// Splits a specific chromosome into chunks.
<a href=#416 id=416 data-nosnippet>416</a>    </span><span class="kw">pub fn </span>chromosome_chunks(<span class="kw-2">&amp;</span><span class="self">self</span>, chrom: <span class="kw-2">&amp;</span>str, bin_size: u64) -&gt; <span class="prelude-ty">Result</span>&lt;Vec&lt;Region&gt;&gt; {
<a href=#417 id=417 data-nosnippet>417</a>        <span class="kw">let </span>genome_chunk_length = <span class="self">self</span>.estimate_genome_chunk_length(bin_size)<span class="question-mark">?</span>;
<a href=#418 id=418 data-nosnippet>418</a>        <span class="kw">let </span>stats = <span class="self">self
<a href=#419 id=419 data-nosnippet>419</a>            </span>.chrom_stats
<a href=#420 id=420 data-nosnippet>420</a>            .get(chrom)
<a href=#421 id=421 data-nosnippet>421</a>            .ok_or_else(|| anyhow::Error::msg(<span class="macro">format!</span>(<span class="string">"Chromosome {chrom} not found"</span>)))<span class="question-mark">?</span>;
<a href=#422 id=422 data-nosnippet>422</a>
<a href=#423 id=423 data-nosnippet>423</a>        <span class="kw">let </span><span class="kw-2">mut </span>chunks = Vec::new();
<a href=#424 id=424 data-nosnippet>424</a>        <span class="kw">let </span><span class="kw-2">mut </span>chunk_start = <span class="number">1</span>;
<a href=#425 id=425 data-nosnippet>425</a>        <span class="kw">let </span>chrom_end = stats.length;
<a href=#426 id=426 data-nosnippet>426</a>
<a href=#427 id=427 data-nosnippet>427</a>        <span class="kw">while </span>chunk_start &lt;= chrom_end {
<a href=#428 id=428 data-nosnippet>428</a>            <span class="comment">// Corrected to include the last position in the range
<a href=#429 id=429 data-nosnippet>429</a>            </span><span class="kw">let </span>chunk_end = chunk_start + genome_chunk_length - <span class="number">1</span>; <span class="comment">// Adjust to ensure the chunk covers exactly genome_chunk_length positions
<a href=#430 id=430 data-nosnippet>430</a>            </span><span class="kw">let </span>chunk_end = chunk_end.min(chrom_end); <span class="comment">// Ensure we do not exceed the chromosome length
<a href=#431 id=431 data-nosnippet>431</a>
<a href=#432 id=432 data-nosnippet>432</a>            </span><span class="kw">let </span>start = Position::try_from(chunk_start <span class="kw">as </span>usize).unwrap();
<a href=#433 id=433 data-nosnippet>433</a>            <span class="kw">let </span>end = Position::try_from(chunk_end <span class="kw">as </span>usize).unwrap();
<a href=#434 id=434 data-nosnippet>434</a>
<a href=#435 id=435 data-nosnippet>435</a>            <span class="kw">let </span>region = Region::new(chrom, start..=end);
<a href=#436 id=436 data-nosnippet>436</a>            chunks.push(region);
<a href=#437 id=437 data-nosnippet>437</a>            chunk_start = chunk_end + <span class="number">1</span>; <span class="comment">// Corrected to start the next chunk right after the current chunk ends
<a href=#438 id=438 data-nosnippet>438</a>        </span>}
<a href=#439 id=439 data-nosnippet>439</a>
<a href=#440 id=440 data-nosnippet>440</a>        <span class="prelude-val">Ok</span>(chunks)
<a href=#441 id=441 data-nosnippet>441</a>    }
<a href=#442 id=442 data-nosnippet>442</a>
<a href=#443 id=443 data-nosnippet>443</a>    <span class="doccomment">/// Returns a mapping from chromosome ID to chromosome name.
<a href=#444 id=444 data-nosnippet>444</a>    </span><span class="kw">pub fn </span>chromosome_id_to_chromosome_name_mapping(<span class="kw-2">&amp;</span><span class="self">self</span>) -&gt; HashMap&lt;usize, String&gt; {
<a href=#445 id=445 data-nosnippet>445</a>        <span class="kw">let </span><span class="kw-2">mut </span>ref_id_mapping = HashMap::default();
<a href=#446 id=446 data-nosnippet>446</a>        <span class="kw">for </span>(i, (name, <span class="kw">_</span>)) <span class="kw">in </span><span class="self">self</span>.header.reference_sequences().iter().enumerate() {
<a href=#447 id=447 data-nosnippet>447</a>            <span class="kw">let </span>name = std::str::from_utf8(name)
<a href=#448 id=448 data-nosnippet>448</a>                .map_err(|e| std::io::Error::new(std::io::ErrorKind::InvalidInput, e))
<a href=#449 id=449 data-nosnippet>449</a>                .unwrap()
<a href=#450 id=450 data-nosnippet>450</a>                .to_string();
<a href=#451 id=451 data-nosnippet>451</a>            ref_id_mapping.insert(i, name);
<a href=#452 id=452 data-nosnippet>452</a>        }
<a href=#453 id=453 data-nosnippet>453</a>        ref_id_mapping
<a href=#454 id=454 data-nosnippet>454</a>    }
<a href=#455 id=455 data-nosnippet>455</a>
<a href=#456 id=456 data-nosnippet>456</a>    <span class="doccomment">/// Returns a mapping from chromosome name to chromosome ID.
<a href=#457 id=457 data-nosnippet>457</a>    </span><span class="kw">pub fn </span>chromosome_name_to_id_mapping(<span class="kw-2">&amp;</span><span class="self">self</span>) -&gt; <span class="prelude-ty">Result</span>&lt;HashMap&lt;String, usize&gt;&gt; {
<a href=#458 id=458 data-nosnippet>458</a>        <span class="kw">let </span><span class="kw-2">mut </span>mapping = HashMap::default();
<a href=#459 id=459 data-nosnippet>459</a>        <span class="kw">let </span>id_to_name = <span class="self">self</span>.chromosome_id_to_chromosome_name_mapping();
<a href=#460 id=460 data-nosnippet>460</a>
<a href=#461 id=461 data-nosnippet>461</a>        <span class="comment">// Invert the mapping
<a href=#462 id=462 data-nosnippet>462</a>        </span><span class="kw">for </span>(id, name) <span class="kw">in </span>id_to_name.iter() {
<a href=#463 id=463 data-nosnippet>463</a>            mapping.insert(name.clone(), <span class="kw-2">*</span>id);
<a href=#464 id=464 data-nosnippet>464</a>        }
<a href=#465 id=465 data-nosnippet>465</a>
<a href=#466 id=466 data-nosnippet>466</a>        <span class="prelude-val">Ok</span>(mapping)
<a href=#467 id=467 data-nosnippet>467</a>    }
<a href=#468 id=468 data-nosnippet>468</a>
<a href=#469 id=469 data-nosnippet>469</a>    <span class="doccomment">/// Returns a mapping from chromosome ID to chromosome length.
<a href=#470 id=470 data-nosnippet>470</a>    </span><span class="kw">pub fn </span>chromsizes_ref_id(<span class="kw-2">&amp;</span><span class="self">self</span>) -&gt; <span class="prelude-ty">Result</span>&lt;HashMap&lt;usize, u64&gt;&gt; {
<a href=#471 id=471 data-nosnippet>471</a>        <span class="kw">let </span><span class="kw-2">mut </span>chromsizes = HashMap::default();
<a href=#472 id=472 data-nosnippet>472</a>        <span class="kw">for </span>(i, (<span class="kw">_</span>, map)) <span class="kw">in </span><span class="self">self</span>.header.reference_sequences().iter().enumerate() {
<a href=#473 id=473 data-nosnippet>473</a>            <span class="kw">let </span>length = map.length().get();
<a href=#474 id=474 data-nosnippet>474</a>            chromsizes.insert(i, length <span class="kw">as </span>u64);
<a href=#475 id=475 data-nosnippet>475</a>        }
<a href=#476 id=476 data-nosnippet>476</a>        <span class="prelude-val">Ok</span>(chromsizes)
<a href=#477 id=477 data-nosnippet>477</a>    }
<a href=#478 id=478 data-nosnippet>478</a>
<a href=#479 id=479 data-nosnippet>479</a>    <span class="doccomment">/// Returns a mapping from chromosome name to chromosome length.
<a href=#480 id=480 data-nosnippet>480</a>    </span><span class="kw">pub fn </span>chromsizes_ref_name(<span class="kw-2">&amp;</span><span class="self">self</span>) -&gt; <span class="prelude-ty">Result</span>&lt;HashMap&lt;String, u64&gt;&gt; {
<a href=#481 id=481 data-nosnippet>481</a>        <span class="kw">let </span><span class="kw-2">mut </span>chromsizes = HashMap::default();
<a href=#482 id=482 data-nosnippet>482</a>        <span class="kw">for </span>(name, map) <span class="kw">in </span><span class="self">self</span>.header.reference_sequences() {
<a href=#483 id=483 data-nosnippet>483</a>            <span class="kw">let </span>length = map.length().get();
<a href=#484 id=484 data-nosnippet>484</a>            <span class="kw">let </span>name = std::str::from_utf8(name)
<a href=#485 id=485 data-nosnippet>485</a>                .map_err(|e| std::io::Error::new(std::io::ErrorKind::InvalidInput, e))
<a href=#486 id=486 data-nosnippet>486</a>                .unwrap()
<a href=#487 id=487 data-nosnippet>487</a>                .to_string();
<a href=#488 id=488 data-nosnippet>488</a>            chromsizes.insert(name, length <span class="kw">as </span>u64);
<a href=#489 id=489 data-nosnippet>489</a>        }
<a href=#490 id=490 data-nosnippet>490</a>        <span class="prelude-val">Ok</span>(chromsizes)
<a href=#491 id=491 data-nosnippet>491</a>    }
<a href=#492 id=492 data-nosnippet>492</a>
<a href=#493 id=493 data-nosnippet>493</a>    <span class="doccomment">/// Writes chromosome sizes to a TSV file.
<a href=#494 id=494 data-nosnippet>494</a>    </span><span class="kw">pub fn </span>write_chromsizes(<span class="kw-2">&amp;</span><span class="self">self</span>, outfile: PathBuf) -&gt; <span class="prelude-ty">Result</span>&lt;()&gt; {
<a href=#495 id=495 data-nosnippet>495</a>        <span class="kw">let </span>chrom_sizes = <span class="self">self</span>.chromsizes_ref_name()<span class="question-mark">?</span>;
<a href=#496 id=496 data-nosnippet>496</a>        <span class="comment">// Convert the chromosome sizes to a DataFrame
<a href=#497 id=497 data-nosnippet>497</a>        </span><span class="kw">let </span><span class="kw-2">mut </span>df = DataFrame::new(<span class="macro">vec!</span>[
<a href=#498 id=498 data-nosnippet>498</a>            Column::new(
<a href=#499 id=499 data-nosnippet>499</a>                <span class="string">"chrom"</span>.into(),
<a href=#500 id=500 data-nosnippet>500</a>                chrom_sizes.keys().cloned().collect::&lt;Vec&lt;String&gt;&gt;(),
<a href=#501 id=501 data-nosnippet>501</a>            ),
<a href=#502 id=502 data-nosnippet>502</a>            Column::new(
<a href=#503 id=503 data-nosnippet>503</a>                <span class="string">"length"</span>.into(),
<a href=#504 id=504 data-nosnippet>504</a>                chrom_sizes.values().cloned().collect::&lt;Vec&lt;u64&gt;&gt;(),
<a href=#505 id=505 data-nosnippet>505</a>            ),
<a href=#506 id=506 data-nosnippet>506</a>        ])<span class="question-mark">?</span>;
<a href=#507 id=507 data-nosnippet>507</a>
<a href=#508 id=508 data-nosnippet>508</a>        <span class="comment">// Sort the DataFrame by chromosome name
<a href=#509 id=509 data-nosnippet>509</a>        </span>df.sort_in_place([<span class="string">"chrom"</span>], SortMultipleOptions::default())<span class="question-mark">?</span>;
<a href=#510 id=510 data-nosnippet>510</a>
<a href=#511 id=511 data-nosnippet>511</a>        <span class="kw">let </span><span class="kw-2">mut </span>file = std::fs::File::create(outfile)<span class="question-mark">?</span>;
<a href=#512 id=512 data-nosnippet>512</a>        CsvWriter::new(<span class="kw-2">&amp;mut </span>file)
<a href=#513 id=513 data-nosnippet>513</a>            .include_header(<span class="bool-val">false</span>)
<a href=#514 id=514 data-nosnippet>514</a>            .with_separator(<span class="string">b'\t'</span>)
<a href=#515 id=515 data-nosnippet>515</a>            .finish(<span class="kw-2">&amp;mut </span>df)<span class="question-mark">?</span>;
<a href=#516 id=516 data-nosnippet>516</a>        <span class="prelude-val">Ok</span>(())
<a href=#517 id=517 data-nosnippet>517</a>    }
<a href=#518 id=518 data-nosnippet>518</a>
<a href=#519 id=519 data-nosnippet>519</a>    <span class="doccomment">/// Returns the number of mapped reads.
<a href=#520 id=520 data-nosnippet>520</a>    </span><span class="kw">pub fn </span>n_mapped(<span class="kw-2">&amp;</span><span class="self">self</span>) -&gt; u64 {
<a href=#521 id=521 data-nosnippet>521</a>        <span class="self">self</span>.n_mapped
<a href=#522 id=522 data-nosnippet>522</a>    }
<a href=#523 id=523 data-nosnippet>523</a>
<a href=#524 id=524 data-nosnippet>524</a>    <span class="doccomment">/// Returns the number of unmapped reads.
<a href=#525 id=525 data-nosnippet>525</a>    </span><span class="kw">pub fn </span>n_unmapped(<span class="kw-2">&amp;</span><span class="self">self</span>) -&gt; u64 {
<a href=#526 id=526 data-nosnippet>526</a>        <span class="self">self</span>.n_unmapped
<a href=#527 id=527 data-nosnippet>527</a>    }
<a href=#528 id=528 data-nosnippet>528</a>
<a href=#529 id=529 data-nosnippet>529</a>    <span class="doccomment">/// Returns the total number of reads.
<a href=#530 id=530 data-nosnippet>530</a>    </span><span class="kw">pub fn </span>n_total_reads(<span class="kw-2">&amp;</span><span class="self">self</span>) -&gt; u64 {
<a href=#531 id=531 data-nosnippet>531</a>        <span class="self">self</span>.n_reads
<a href=#532 id=532 data-nosnippet>532</a>    }
<a href=#533 id=533 data-nosnippet>533</a>}
<a href=#534 id=534 data-nosnippet>534</a>
<a href=#535 id=535 data-nosnippet>535</a><span class="doccomment">/// Supported output file types.
<a href=#536 id=536 data-nosnippet>536</a></span><span class="kw">pub enum </span>FileType {
<a href=#537 id=537 data-nosnippet>537</a>    <span class="doccomment">/// BedGraph format.
<a href=#538 id=538 data-nosnippet>538</a>    </span>Bedgraph,
<a href=#539 id=539 data-nosnippet>539</a>    <span class="doccomment">/// BigWig format.
<a href=#540 id=540 data-nosnippet>540</a>    </span>Bigwig,
<a href=#541 id=541 data-nosnippet>541</a>    <span class="doccomment">/// Tab-Separated Values.
<a href=#542 id=542 data-nosnippet>542</a>    </span>TSV,
<a href=#543 id=543 data-nosnippet>543</a>}
<a href=#544 id=544 data-nosnippet>544</a>
<a href=#545 id=545 data-nosnippet>545</a><span class="kw">impl </span>std::str::FromStr <span class="kw">for </span>FileType {
<a href=#546 id=546 data-nosnippet>546</a>    <span class="kw">type </span><span class="prelude-val">Err </span>= String;
<a href=#547 id=547 data-nosnippet>547</a>
<a href=#548 id=548 data-nosnippet>548</a>    <span class="kw">fn </span>from_str(s: <span class="kw-2">&amp;</span>str) -&gt; <span class="prelude-ty">Result</span>&lt;FileType, String&gt; {
<a href=#549 id=549 data-nosnippet>549</a>        <span class="kw">match </span>s.to_lowercase().as_str() {
<a href=#550 id=550 data-nosnippet>550</a>            <span class="string">"bedgraph" </span>=&gt; <span class="prelude-val">Ok</span>(FileType::Bedgraph),
<a href=#551 id=551 data-nosnippet>551</a>            <span class="string">"bdg" </span>=&gt; <span class="prelude-val">Ok</span>(FileType::Bedgraph),
<a href=#552 id=552 data-nosnippet>552</a>            <span class="string">"bigwig" </span>=&gt; <span class="prelude-val">Ok</span>(FileType::Bigwig),
<a href=#553 id=553 data-nosnippet>553</a>            <span class="string">"bw" </span>=&gt; <span class="prelude-val">Ok</span>(FileType::Bigwig),
<a href=#554 id=554 data-nosnippet>554</a>            <span class="string">"tsv" </span>=&gt; <span class="prelude-val">Ok</span>(FileType::TSV),
<a href=#555 id=555 data-nosnippet>555</a>            <span class="string">"txt" </span>=&gt; <span class="prelude-val">Ok</span>(FileType::TSV),
<a href=#556 id=556 data-nosnippet>556</a>            <span class="kw">_ </span>=&gt; <span class="prelude-val">Err</span>(<span class="macro">format!</span>(<span class="string">"Unknown file type: {s}"</span>)),
<a href=#557 id=557 data-nosnippet>557</a>        }
<a href=#558 id=558 data-nosnippet>558</a>    }
<a href=#559 id=559 data-nosnippet>559</a>}
<a href=#560 id=560 data-nosnippet>560</a>
<a href=#561 id=561 data-nosnippet>561</a><span class="doccomment">/// Converts a list of regions to a `Lapper` interval tree for efficient overlap queries.
<a href=#562 id=562 data-nosnippet>562</a></span><span class="kw">pub fn </span>regions_to_lapper(regions: Vec&lt;Region&gt;) -&gt; <span class="prelude-ty">Result</span>&lt;HashMap&lt;String, Lapper&lt;usize, u32&gt;&gt;&gt; {
<a href=#563 id=563 data-nosnippet>563</a>    <span class="kw">let </span><span class="kw-2">mut </span>lapper: HashMap&lt;String, Lapper&lt;usize, u32&gt;&gt; = HashMap::default();
<a href=#564 id=564 data-nosnippet>564</a>    <span class="kw">let </span><span class="kw-2">mut </span>intervals: HashMap&lt;String, Vec&lt;Iv&gt;&gt; = HashMap::default();
<a href=#565 id=565 data-nosnippet>565</a>
<a href=#566 id=566 data-nosnippet>566</a>    <span class="kw">for </span>reg <span class="kw">in </span>regions {
<a href=#567 id=567 data-nosnippet>567</a>        <span class="kw">let </span>chrom = reg.name().to_string();
<a href=#568 id=568 data-nosnippet>568</a>        <span class="kw">let </span>start = reg.start().map(|x| x.get());
<a href=#569 id=569 data-nosnippet>569</a>        <span class="kw">let </span>end = reg.end().map(|x| x.get());
<a href=#570 id=570 data-nosnippet>570</a>
<a href=#571 id=571 data-nosnippet>571</a>        <span class="kw">let </span>start = <span class="kw">match </span>start {
<a href=#572 id=572 data-nosnippet>572</a>            Bound::Included(start) =&gt; start,
<a href=#573 id=573 data-nosnippet>573</a>            Bound::Excluded(start) =&gt; start + <span class="number">1</span>,
<a href=#574 id=574 data-nosnippet>574</a>            <span class="kw">_ </span>=&gt; <span class="number">0</span>,
<a href=#575 id=575 data-nosnippet>575</a>        };
<a href=#576 id=576 data-nosnippet>576</a>
<a href=#577 id=577 data-nosnippet>577</a>        <span class="kw">let </span>end = <span class="kw">match </span>end {
<a href=#578 id=578 data-nosnippet>578</a>            Bound::Included(end) =&gt; end,
<a href=#579 id=579 data-nosnippet>579</a>            Bound::Excluded(end) =&gt; end - <span class="number">1</span>,
<a href=#580 id=580 data-nosnippet>580</a>            <span class="kw">_ </span>=&gt; <span class="number">0</span>,
<a href=#581 id=581 data-nosnippet>581</a>        };
<a href=#582 id=582 data-nosnippet>582</a>
<a href=#583 id=583 data-nosnippet>583</a>        <span class="kw">let </span>iv = Iv {
<a href=#584 id=584 data-nosnippet>584</a>            start,
<a href=#585 id=585 data-nosnippet>585</a>            stop: end,
<a href=#586 id=586 data-nosnippet>586</a>            val: <span class="number">0</span>,
<a href=#587 id=587 data-nosnippet>587</a>        };
<a href=#588 id=588 data-nosnippet>588</a>        intervals.entry(chrom.clone()).or_default().push(iv);
<a href=#589 id=589 data-nosnippet>589</a>    }
<a href=#590 id=590 data-nosnippet>590</a>
<a href=#591 id=591 data-nosnippet>591</a>    <span class="kw">for </span>(chrom, ivs) <span class="kw">in </span>intervals.iter() {
<a href=#592 id=592 data-nosnippet>592</a>        <span class="kw">let </span>lap = Lapper::new(ivs.to_vec());
<a href=#593 id=593 data-nosnippet>593</a>        lapper.insert(chrom.clone(), lap);
<a href=#594 id=594 data-nosnippet>594</a>    }
<a href=#595 id=595 data-nosnippet>595</a>
<a href=#596 id=596 data-nosnippet>596</a>    <span class="prelude-val">Ok</span>(lapper)
<a href=#597 id=597 data-nosnippet>597</a>}
<a href=#598 id=598 data-nosnippet>598</a>
<a href=#599 id=599 data-nosnippet>599</a><span class="doccomment">/// Reads a BED file and converts it to a `Lapper` interval tree.
<a href=#600 id=600 data-nosnippet>600</a></span><span class="kw">pub fn </span>bed_to_lapper(bed: PathBuf) -&gt; <span class="prelude-ty">Result</span>&lt;HashMap&lt;String, Lapper&lt;usize, u32&gt;&gt;&gt; {
<a href=#601 id=601 data-nosnippet>601</a>    <span class="comment">// Read the bed file and convert it to a lapper
<a href=#602 id=602 data-nosnippet>602</a>    </span><span class="kw">let </span>reader = std::fs::File::open(bed)<span class="question-mark">?</span>;
<a href=#603 id=603 data-nosnippet>603</a>    <span class="kw">let </span>buf_reader = std::io::BufReader::new(reader);
<a href=#604 id=604 data-nosnippet>604</a>    <span class="kw">let </span><span class="kw-2">mut </span>bed_reader = bed::io::Reader::&lt;<span class="number">4</span>, <span class="kw">_</span>&gt;::new(buf_reader);
<a href=#605 id=605 data-nosnippet>605</a>    <span class="kw">let </span><span class="kw-2">mut </span>record = bed::Record::default();
<a href=#606 id=606 data-nosnippet>606</a>    <span class="kw">let </span><span class="kw-2">mut </span>intervals: HashMap&lt;String, Vec&lt;Iv&gt;&gt; = HashMap::default();
<a href=#607 id=607 data-nosnippet>607</a>    <span class="kw">let </span><span class="kw-2">mut </span>lapper: HashMap&lt;String, Lapper&lt;usize, u32&gt;&gt; = HashMap::default();
<a href=#608 id=608 data-nosnippet>608</a>
<a href=#609 id=609 data-nosnippet>609</a>    <span class="kw">while </span>bed_reader.read_record(<span class="kw-2">&amp;mut </span>record)<span class="question-mark">? </span>!= <span class="number">0 </span>{
<a href=#610 id=610 data-nosnippet>610</a>        <span class="kw">let </span>chrom = record.reference_sequence_name().to_string();
<a href=#611 id=611 data-nosnippet>611</a>        <span class="kw">let </span>start = record.feature_start()<span class="question-mark">?</span>;
<a href=#612 id=612 data-nosnippet>612</a>        <span class="kw">let </span>end = record
<a href=#613 id=613 data-nosnippet>613</a>            .feature_end()
<a href=#614 id=614 data-nosnippet>614</a>            .context(<span class="string">"Failed to get feature end"</span>)<span class="question-mark">??</span>;
<a href=#615 id=615 data-nosnippet>615</a>
<a href=#616 id=616 data-nosnippet>616</a>        <span class="kw">let </span>iv = Iv {
<a href=#617 id=617 data-nosnippet>617</a>            start: start.get(),
<a href=#618 id=618 data-nosnippet>618</a>            stop: end.get(),
<a href=#619 id=619 data-nosnippet>619</a>            val: <span class="number">0</span>,
<a href=#620 id=620 data-nosnippet>620</a>        };
<a href=#621 id=621 data-nosnippet>621</a>        intervals.entry(chrom.clone()).or_default().push(iv);
<a href=#622 id=622 data-nosnippet>622</a>    }
<a href=#623 id=623 data-nosnippet>623</a>
<a href=#624 id=624 data-nosnippet>624</a>    <span class="kw">for </span>(chrom, ivs) <span class="kw">in </span>intervals.iter() {
<a href=#625 id=625 data-nosnippet>625</a>        <span class="kw">let </span>lap = Lapper::new(ivs.to_vec());
<a href=#626 id=626 data-nosnippet>626</a>        lapper.insert(chrom.clone(), lap);
<a href=#627 id=627 data-nosnippet>627</a>    }
<a href=#628 id=628 data-nosnippet>628</a>    <span class="comment">// Check if the lapper is empty
<a href=#629 id=629 data-nosnippet>629</a>    </span><span class="kw">if </span>lapper.is_empty() {
<a href=#630 id=630 data-nosnippet>630</a>        <span class="kw">return </span><span class="prelude-val">Err</span>(anyhow::Error::msg(<span class="string">"Lapper is empty"</span>));
<a href=#631 id=631 data-nosnippet>631</a>    }
<a href=#632 id=632 data-nosnippet>632</a>
<a href=#633 id=633 data-nosnippet>633</a>    <span class="prelude-val">Ok</span>(lapper)
<a href=#634 id=634 data-nosnippet>634</a>}
<a href=#635 id=635 data-nosnippet>635</a>
<a href=#636 id=636 data-nosnippet>636</a><span class="doccomment">/// Converts the chromosome names in a `Lapper` to chromosome IDs based on a BAM file's header.
<a href=#637 id=637 data-nosnippet>637</a>///
<a href=#638 id=638 data-nosnippet>638</a>/// This is useful for when we want to use the lapper with the BAM file.
<a href=#639 id=639 data-nosnippet>639</a></span><span class="kw">pub fn </span>convert_lapper_chrom_names_to_ids(
<a href=#640 id=640 data-nosnippet>640</a>    lapper: HashMap&lt;String, Lapper&lt;usize, u32&gt;&gt;,
<a href=#641 id=641 data-nosnippet>641</a>    bam_stats: <span class="kw-2">&amp;</span>BamStats,
<a href=#642 id=642 data-nosnippet>642</a>) -&gt; <span class="prelude-ty">Result</span>&lt;HashMap&lt;usize, Lapper&lt;usize, u32&gt;&gt;&gt; {
<a href=#643 id=643 data-nosnippet>643</a>    <span class="comment">// Convert the chromosome names in the lapper to the chromosome IDs in the BAM file
<a href=#644 id=644 data-nosnippet>644</a>    </span><span class="kw">let </span>chrom_id_mapping = bam_stats.chromosome_name_to_id_mapping()<span class="question-mark">?</span>;
<a href=#645 id=645 data-nosnippet>645</a>    <span class="kw">let </span><span class="kw-2">mut </span>lapper_chrom_id: HashMap&lt;usize, Lapper&lt;usize, u32&gt;&gt; = HashMap::default();
<a href=#646 id=646 data-nosnippet>646</a>    <span class="kw">for </span>(chrom, lap) <span class="kw">in </span>lapper.iter() {
<a href=#647 id=647 data-nosnippet>647</a>        <span class="kw">let </span>chrom_id = chrom_id_mapping.get(chrom).ok_or_else(|| {
<a href=#648 id=648 data-nosnippet>648</a>            anyhow::Error::msg(<span class="macro">format!</span>(<span class="string">"Chromosome {chrom} not found in BAM file"</span>))
<a href=#649 id=649 data-nosnippet>649</a>        })<span class="question-mark">?</span>;
<a href=#650 id=650 data-nosnippet>650</a>        lapper_chrom_id.insert(<span class="kw-2">*</span>chrom_id, lap.clone());
<a href=#651 id=651 data-nosnippet>651</a>    }
<a href=#652 id=652 data-nosnippet>652</a>    <span class="prelude-val">Ok</span>(lapper_chrom_id)
<a href=#653 id=653 data-nosnippet>653</a>}
<a href=#654 id=654 data-nosnippet>654</a>
<a href=#655 id=655 data-nosnippet>655</a><span class="doccomment">/// Reads the header of a BAM file.
<a href=#656 id=656 data-nosnippet>656</a>///
<a href=#657 id=657 data-nosnippet>657</a>/// If the `noodles` crate fails to read the header, it falls back to `samtools`.
<a href=#658 id=658 data-nosnippet>658</a>/// This is useful for handling BAM files with slightly non-standard headers (e.g., from CellRanger).
<a href=#659 id=659 data-nosnippet>659</a></span><span class="kw">pub fn </span>get_bam_header&lt;P&gt;(file_path: P) -&gt; <span class="prelude-ty">Result</span>&lt;sam::Header&gt;
<a href=#660 id=660 data-nosnippet>660</a><span class="kw">where
<a href=#661 id=661 data-nosnippet>661</a>    </span>P: AsRef&lt;Path&gt;,
<a href=#662 id=662 data-nosnippet>662</a>{
<a href=#663 id=663 data-nosnippet>663</a>    <span class="kw">let </span>file_path = file_path.as_ref();
<a href=#664 id=664 data-nosnippet>664</a>
<a href=#665 id=665 data-nosnippet>665</a>    <span class="comment">// Check that the file exists
<a href=#666 id=666 data-nosnippet>666</a>    </span><span class="kw">if </span>!file_path.exists() {
<a href=#667 id=667 data-nosnippet>667</a>        <span class="kw">return </span><span class="prelude-val">Err</span>(anyhow::Error::from(std::io::Error::new(
<a href=#668 id=668 data-nosnippet>668</a>            std::io::ErrorKind::NotFound,
<a href=#669 id=669 data-nosnippet>669</a>            <span class="macro">format!</span>(<span class="string">"File not found: {}"</span>, file_path.display()),
<a href=#670 id=670 data-nosnippet>670</a>        )));
<a href=#671 id=671 data-nosnippet>671</a>    };
<a href=#672 id=672 data-nosnippet>672</a>
<a href=#673 id=673 data-nosnippet>673</a>    <span class="kw">let </span><span class="kw-2">mut </span>reader = bam::io::indexed_reader::Builder::default()
<a href=#674 id=674 data-nosnippet>674</a>        .build_from_path(file_path)
<a href=#675 id=675 data-nosnippet>675</a>        .expect(<span class="string">"Failed to open file"</span>);
<a href=#676 id=676 data-nosnippet>676</a>
<a href=#677 id=677 data-nosnippet>677</a>    <span class="kw">let </span>header = <span class="kw">match </span>reader.read_header() {
<a href=#678 id=678 data-nosnippet>678</a>        std::result::Result::Ok(header) =&gt; header,
<a href=#679 id=679 data-nosnippet>679</a>        <span class="prelude-val">Err</span>(e) =&gt; {
<a href=#680 id=680 data-nosnippet>680</a>            <span class="macro">debug!</span>(<span class="string">"Failed to read header using noodels falling back to samtools: {e}"</span>);
<a href=#681 id=681 data-nosnippet>681</a>
<a href=#682 id=682 data-nosnippet>682</a>            <span class="kw">let </span>header_samtools = std::process::Command::new(<span class="string">"samtools"</span>)
<a href=#683 id=683 data-nosnippet>683</a>                .arg(<span class="string">"view"</span>)
<a href=#684 id=684 data-nosnippet>684</a>                .arg(<span class="string">"-H"</span>)
<a href=#685 id=685 data-nosnippet>685</a>                .arg(file_path)
<a href=#686 id=686 data-nosnippet>686</a>                .output()
<a href=#687 id=687 data-nosnippet>687</a>                .expect(<span class="string">"Failed to run samtools"</span>)
<a href=#688 id=688 data-nosnippet>688</a>                .stdout;
<a href=#689 id=689 data-nosnippet>689</a>
<a href=#690 id=690 data-nosnippet>690</a>            <span class="kw">let </span>header_str =
<a href=#691 id=691 data-nosnippet>691</a>                String::from_utf8(header_samtools).expect(<span class="string">"Failed to convert header to string"</span>);
<a href=#692 id=692 data-nosnippet>692</a>
<a href=#693 id=693 data-nosnippet>693</a>            <span class="comment">// Slight hack here for CellRanger BAM files that are missing the version info
<a href=#694 id=694 data-nosnippet>694</a>            </span><span class="kw">let </span>header_string =
<a href=#695 id=695 data-nosnippet>695</a>                header_str.replace(<span class="string">"@HD\tSO:coordinate\n"</span>, <span class="string">"@HD\tVN:1.6\tSO:coordinate\n"</span>);
<a href=#696 id=696 data-nosnippet>696</a>            <span class="kw">let </span>header_str = header_string.as_bytes();
<a href=#697 id=697 data-nosnippet>697</a>            <span class="kw">let </span><span class="kw-2">mut </span>reader = sam::io::Reader::new(header_str);
<a href=#698 id=698 data-nosnippet>698</a>            reader
<a href=#699 id=699 data-nosnippet>699</a>                .read_header()
<a href=#700 id=700 data-nosnippet>700</a>                .expect(<span class="string">"Failed to read header with samtools"</span>)
<a href=#701 id=701 data-nosnippet>701</a>        }
<a href=#702 id=702 data-nosnippet>702</a>    };
<a href=#703 id=703 data-nosnippet>703</a>    <span class="prelude-val">Ok</span>(header)
<a href=#704 id=704 data-nosnippet>704</a>}</code></pre></div></section></main></body></html>