rd_("AjThe cell barcode tag (CB).CiType alias for an interval with a <code>u32</code> value.B`Counts Per Million mapped reads.AaNo normalization.AeTab-Separated Values.BgCreates a new <code>BamModifier</code>.BgCreates a new <code>BamFilterer</code>.CiCreates a new empty <code>CellBarcodes</code> collection.CnCreates a new empty <code>CellBarcodesMulti</code> collection.DbCreates a new <code>BamStats</code> instance from a BAM file path.BdCreate a new <code>BamPileup</code>.EbCreate a new <code>MultiBamPileup</code> from a vector of <code>BamPileup</code>s.BiCreates a new <code>IntervalMaker</code>.DjCreates a new <code>BamReadFilterStats</code> with all counters set to \xe2\x80\xa6BiCreates a new <code>BamReadFilter</code>.BfCreate a new <code>BamSplitter</code>.BmRuns the modification process asynchronously.AjBoth strands (unstranded).BlReads Per Kilobase per Million mapped reads.AoReturns the argument unchanged.0000000000000000000BaCalls <code>U::from(self)</code>.0000000000000000000CdRepresents coordinate shifts to be applied to reads.CnSplit the input BAM file into endogenous and exogenous BAM \xe2\x80\xa6CdReturns a snapshot of the current filter statistics.BaGet the statistics for the split.nBigWig format.BkRepresents the strand of a genomic feature.BkCalculates the coordinates of the interval.CcCreates a new BAM record with modified coordinates.BjWrites the multi-BAM pileup to a TSV file.AcForward strand (+).AcReverse strand (-).oBamNado LibraryC`Stores statistics and metadata about a BAM file.A`BedGraph format.AlSupported output file types.BgRepresents truncation values for reads.AhBamNado Python InterfaceBfReturns a copy of the set of barcodes.BmReturns a copy of the vector of barcode sets.BdLoads cell barcodes from a CSV file.C`Loads multiple sets of barcodes from a CSV file.CgParse a string into a <code>NormalizationMethod</code>.CeReturns <code>true</code> if the collection is empty.0CkChecks if a read is valid according to the filter criteria.BcReturns the number of mapped reads.BkTakes a snapshot of the current statistics.AdBAM Utilities ModuleClGet the path to the BAM file containing reads mapping to \xe2\x80\xa6BmWrite the normalized pileup as a BigWig file.BjStatistics for the read splitting process.AoReads the header of a BAM file.BhShift for the 5\xe2\x80\x99 end of forward reads.BeReturns the number of unmapped reads.BiFilters BAM files based on read criteria.ChModifies BAM files by filtering reads and optionally \xe2\x80\xa6CgA tool for splitting a BAM file into endogenous and \xe2\x80\xa6AbRead Filter ModuleBjRuns the filtering process asynchronously.BhShift for the 3\xe2\x80\x99 end of forward reads.BoWrite the normalized pileup as a bedGraph file.BeA collection of unique cell barcodes.AcBAM Modifier ModuleAcBAM Splitter ModuleBmCreates a progress bar with a standard style.ClCalculate the scaling factor for the given normalization \xe2\x80\xa6AgA filter for BAM reads.CiHelper struct for creating genomic intervals from BAM \xe2\x80\xa6DhReads a BED file and converts it to a <code>Lapper</code> interval tree.CfSplits the genome into chunks for parallel processing.BbReturns the total number of reads.Ck<code>ToString::to_string</code>, but without panic on OOM.0000CgGet the path to the BAM file containing unmapped reads.CkA struct for processing multiple BAM files in parallel. \xe2\x80\xa6ChBigWig file comparison utilities. Aimed at comparing \xe2\x80\xa6BgGet the path to the exogenous BAM file.AoReads the header of a BAM file.CkCompare two BigWig files and output the result as a BigWig.BhGet the path to the endogenous BAM file.BfWrites chromosome sizes to a TSV file.CiA collection of sets of cell barcodes, where each set \xe2\x80\xa6CgAggregate multiple BigWigs into a single output BigWig.BiSplits a specific chromosome into chunks.CjReturns a mapping from chromosome ID to chromosome length.AhCoverage Analysis ModuleAhGenomic Intervals ModuleCmGenerate pileup for a single chromosome. Returns a vector \xe2\x80\xa6DiConverts a list of regions to a <code>Lapper</code> interval tree for \xe2\x80\xa6AhSpike-in Analysis ModuleBjStatistics for the read filtering process.BhShift for the 5\xe2\x80\x99 end of reverse reads.CcNormalization methods for genomic coverage signals.ChReturns a mapping from chromosome name to chromosome \xe2\x80\xa6BhShift for the 3\xe2\x80\x99 end of reverse reads.AkSignal Normalization ModuleDeCollapse <em>adjacent</em> bins (contiguous end==next start) with \xe2\x80\xa6CfReturns the number of reads remaining after filtering.CjA snapshot of the read filter statistics at a specific \xe2\x80\xa6CjEstimates the optimal genome chunk length for parallel \xe2\x80\xa6ChReturns a mapping from chromosome name to chromosome ID.DhConverts the chromosome names in a <code>Lapper</code> to chromosome \xe2\x80\xa6ChReturns a mapping from chromosome ID to chromosome name.")